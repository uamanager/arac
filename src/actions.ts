export enum Actions {
  CREATE = 'create',
  READ = 'read',
  UPDATE = 'update',
  DELETE = 'delete'
}
