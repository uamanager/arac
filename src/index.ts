export * from './accessControl';
export * from './accessCreate';
export * from './accessRequest';
export * from './actions';
export * from './dynamicChecker';
export * from './permission';
export * from './resource';
export * from './role';
export * from './staticChecker';
