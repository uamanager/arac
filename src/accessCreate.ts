import {AccessControl} from './accessControl';
import {Actions} from './actions';
import {TDynamicCheckerFunction} from './dynamicChecker';

export class AccessCreate {
  constructor (
    public roleName: string,
    public allowed: boolean,
    public accessControl: AccessControl
  ) {}

  public create (resourceName: string, dynamicCheckers?: {
    [key: string]: TDynamicCheckerFunction
  }) {
    this.accessControl.permission(
      Actions.CREATE,
      AccessControl.hash(this.roleName, resourceName),
      this.allowed,
      dynamicCheckers
    );

    return this;
  }

  public read (resourceName: string, dynamicCheckers?: {
    [key: string]: TDynamicCheckerFunction
  }) {
    this.accessControl.permission(
      Actions.READ,
      AccessControl.hash(this.roleName, resourceName),
      this.allowed,
      dynamicCheckers
    );

    return this;
  }

  public update (resourceName: string, dynamicCheckers?: {
    [key: string]: TDynamicCheckerFunction
  }) {
    this.accessControl.permission(
      Actions.UPDATE,
      AccessControl.hash(this.roleName, resourceName),
      this.allowed,
      dynamicCheckers
    );

    return this;
  }

  public delete (resourceName: string, dynamicCheckers?: {
    [key: string]: TDynamicCheckerFunction
  }) {
    this.accessControl.permission(
      Actions.DELETE,
      AccessControl.hash(this.roleName, resourceName),
      this.allowed,
      dynamicCheckers
    );

    return this;
  }
}
