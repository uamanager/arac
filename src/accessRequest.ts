import {AccessControl} from './accessControl';
import {Actions} from './actions';
import {Permission} from './permission';

export class AccessRequest {
  constructor (
    public roleName: string,
    public accessControl: AccessControl
  ) {}

  create (resourceName: string) {
    const permissions = this.accessControl.permissions[AccessControl.hash(
      this.roleName,
      resourceName
    )];
    return this.check(Actions.CREATE, resourceName, permissions);
  }

  read (resourceName: string) {
    const permissions = this.accessControl.permissions[AccessControl.hash(
      this.roleName,
      resourceName
    )];
    return this.check(Actions.READ, resourceName, permissions);
  }

  update (resourceName: string) {
    const permissions = this.accessControl.permissions[AccessControl.hash(
      this.roleName,
      resourceName
    )];
    return this.check(Actions.UPDATE, resourceName, permissions);
  }

  delete (resourceName: string) {
    const permissions = this.accessControl.permissions[AccessControl.hash(
      this.roleName,
      resourceName
    )];
    return this.check(Actions.DELETE, resourceName, permissions);
  }

  private check (action: Actions, resourceName: string, permissions: Permission) {
    const dynamicCheckers = permissions.dynamicCheckers[action];
    const dynamicCheckersArray = Object.keys(dynamicCheckers)
      .map(name => dynamicCheckers[name]);
    const checkers = [permissions.staticChecker, ...dynamicCheckersArray];

    return checkers.reduce((prev, next) => {
      return prev.then(() => !!next.check ? next.check(
        action,
        this.roleName,
        resourceName
      ) : Promise.resolve())
        .catch((err) => {
          return Promise.reject(err);
        });
    }, Promise.resolve(true));
  }
}
