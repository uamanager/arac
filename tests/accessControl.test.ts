import {AccessControl} from '../src';
import {checkersMap} from './data/checkers';
import * as dump from './data/dump.json';
import * as simpleDump from './data/simplified-dump.json';
import * as errorStaticDump from './data/error-static-dump.json';
import * as errorActionDump from './data/error-action-dump.json';

let arac;

beforeEach(() => {
  arac = new AccessControl();
});

afterEach(() => {
  arac = undefined;
});


describe('Access Control', () => {

  it('creates instance of AccessControl', () => {
    expect(arac instanceof AccessControl).toBe(true);
  });

  it('imports dump', () => {
    arac.import(dump, checkersMap);

    expect(arac).toMatchSnapshot();
  });

  it('imports simplified dump', () => {
    arac.import(simpleDump, checkersMap);

    expect(arac).toMatchSnapshot();
  });

  it('throws error if checkersMap is not specified but used', () => {
    expect(() => {arac.import(simpleDump, {});}).toThrowError(Error);
  });

  it('exports dump', () => {
    arac.import(dump, checkersMap);
    const _dump = arac.export();

    _dump.forEach(permission => {
      return expect(dump).toEqual(expect.arrayContaining([
        expect.objectContaining(permission)
      ]));
    });

    dump.forEach(permission => {
      return expect(_dump).toEqual(expect.arrayContaining([
        expect.objectContaining(permission)
      ]));
    });

    expect(_dump.length).toBe(dump.length);
  });

  it('returns roles', () => {
    arac.import(simpleDump, checkersMap);

    expect(arac.roles).toMatchSnapshot();
  });

  it('returns resources', () => {
    arac.import(simpleDump, checkersMap);

    expect(arac.resources).toMatchSnapshot();
  });

  it('creates role and not creates role copy or throws error', () => {
    arac.role('testRole');
    expect(arac).toMatchSnapshot();

    arac.role('testRole');
    expect(arac).toMatchSnapshot();
  });

  it('creates first level resource', () => {
    arac.resource('testResource1');
    expect(arac).toMatchSnapshot();
  });

  it('creates second level resource', () => {
    arac.resource('testResource1/testResource2');
    expect(arac).toMatchSnapshot();
  });

  it('creates roles and resources in chain', () => {
    arac
      .role('testRole1')
      .resource('testResource1')
      .resource('testResource2')
      .role('testRole2')
      .role('testRole3');
    expect(arac).toMatchSnapshot();
  });

  it('creates resources on existing level', () => {
    arac
      .resource('testResource1')
      .resource('testResource1/testResource2')
      .resource('testResource1/testResource3');
    expect(arac).toMatchSnapshot();
  });

  it('returns existing resources', () => {
    const _old = arac.resource('testResource1/testResource3');
    const _new = arac.resource('testResource1/testResource3');
    expect(_new.path).toBe(_old.path);
  });

  it('creates root resource', () => {
    arac.resource('/');
    expect(arac).toMatchSnapshot();
  });

  it('throws exception for dynamicChecker with name "static" while import', () => {
    expect(() => arac.import(errorStaticDump, checkersMap)).toThrowError(Error);
  });

  it('throws exception for permission with incorrect action while import', () => {
    expect(() => arac.import(errorActionDump, checkersMap)).toThrowError(Error);
  });
});
