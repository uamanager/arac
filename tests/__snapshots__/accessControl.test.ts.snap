// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Access Control creates first level resource 1`] = `
AccessControl {
  "_permissions": Object {},
  "_resources": Resource {
    "accessControl": [Circular],
    "children": Object {
      "testResource1": Resource {
        "accessControl": [Circular],
        "children": Object {},
        "name": "testResource1",
        "parent": [Circular],
        "path": "testResource1",
      },
    },
    "name": "",
    "parent": null,
    "path": "",
  },
  "_roles": Object {},
}
`;

exports[`Access Control creates resources on existing level 1`] = `
AccessControl {
  "_permissions": Object {},
  "_resources": Resource {
    "accessControl": [Circular],
    "children": Object {
      "testResource1": Resource {
        "accessControl": [Circular],
        "children": Object {
          "testResource2": Resource {
            "accessControl": [Circular],
            "children": Object {},
            "name": "testResource2",
            "parent": [Circular],
            "path": "testResource1/testResource2",
          },
          "testResource3": Resource {
            "accessControl": [Circular],
            "children": Object {},
            "name": "testResource3",
            "parent": [Circular],
            "path": "testResource1/testResource3",
          },
        },
        "name": "testResource1",
        "parent": [Circular],
        "path": "testResource1",
      },
    },
    "name": "",
    "parent": null,
    "path": "",
  },
  "_roles": Object {},
}
`;

exports[`Access Control creates role and not creates role copy or throws error 1`] = `
AccessControl {
  "_permissions": Object {},
  "_resources": Resource {
    "accessControl": [Circular],
    "children": Object {},
    "name": "",
    "parent": null,
    "path": "",
  },
  "_roles": Object {
    "testRole": Role {
      "accessControl": [Circular],
      "name": "testRole",
    },
  },
}
`;

exports[`Access Control creates role and not creates role copy or throws error 2`] = `
AccessControl {
  "_permissions": Object {},
  "_resources": Resource {
    "accessControl": [Circular],
    "children": Object {},
    "name": "",
    "parent": null,
    "path": "",
  },
  "_roles": Object {
    "testRole": Role {
      "accessControl": [Circular],
      "name": "testRole",
    },
  },
}
`;

exports[`Access Control creates roles and resources in chain 1`] = `
AccessControl {
  "_permissions": Object {},
  "_resources": Resource {
    "accessControl": [Circular],
    "children": Object {
      "testResource1": Resource {
        "accessControl": [Circular],
        "children": Object {},
        "name": "testResource1",
        "parent": [Circular],
        "path": "testResource1",
      },
      "testResource2": Resource {
        "accessControl": [Circular],
        "children": Object {},
        "name": "testResource2",
        "parent": [Circular],
        "path": "testResource2",
      },
    },
    "name": "",
    "parent": null,
    "path": "",
  },
  "_roles": Object {
    "testRole1": Role {
      "accessControl": [Circular],
      "name": "testRole1",
    },
    "testRole2": Role {
      "accessControl": [Circular],
      "name": "testRole2",
    },
    "testRole3": Role {
      "accessControl": [Circular],
      "name": "testRole3",
    },
  },
}
`;

exports[`Access Control creates root resource 1`] = `
AccessControl {
  "_permissions": Object {},
  "_resources": Resource {
    "accessControl": [Circular],
    "children": Object {
      "": Resource {
        "accessControl": [Circular],
        "children": Object {},
        "name": "",
        "parent": [Circular],
        "path": "",
      },
    },
    "name": "",
    "parent": null,
    "path": "",
  },
  "_roles": Object {},
}
`;

exports[`Access Control creates second level resource 1`] = `
AccessControl {
  "_permissions": Object {},
  "_resources": Resource {
    "accessControl": [Circular],
    "children": Object {
      "testResource1": Resource {
        "accessControl": [Circular],
        "children": Object {
          "testResource2": Resource {
            "accessControl": [Circular],
            "children": Object {},
            "name": "testResource2",
            "parent": [Circular],
            "path": "testResource1/testResource2",
          },
        },
        "name": "testResource1",
        "parent": [Circular],
        "path": "testResource1",
      },
    },
    "name": "",
    "parent": null,
    "path": "",
  },
  "_roles": Object {},
}
`;

exports[`Access Control imports dump 1`] = `
AccessControl {
  "_permissions": Object {
    "resource1#role1": Permission {
      "dynamicCheckers": Object {
        "create": Object {
          "dynamicCheckerResolve": DynamicChecker {
            "checker": [MockFunction],
          },
        },
        "delete": Object {
          "dynamicCheckerReject": DynamicChecker {
            "checker": [MockFunction],
          },
        },
        "read": Object {
          "dynamicCheckerResolve": DynamicChecker {
            "checker": [MockFunction],
          },
        },
        "update": Object {
          "dynamicCheckerReject": DynamicChecker {
            "checker": [MockFunction],
          },
        },
      },
      "resource": Resource {
        "accessControl": [Circular],
        "children": Object {},
        "name": "resource1",
        "parent": Resource {
          "accessControl": [Circular],
          "children": Object {
            "resource1": [Circular],
            "resource2": Resource {
              "accessControl": [Circular],
              "children": Object {},
              "name": "resource2",
              "parent": [Circular],
              "path": "resource2",
            },
          },
          "name": "",
          "parent": null,
          "path": "",
        },
        "path": "resource1",
      },
      "role": Role {
        "accessControl": [Circular],
        "name": "role1",
      },
      "staticChecker": StaticChecker {
        "permissions": Object {
          "create": true,
          "delete": false,
          "read": false,
          "update": true,
        },
      },
    },
    "resource2#role1": Permission {
      "dynamicCheckers": Object {
        "create": Object {
          "dynamicCheckerResolve": DynamicChecker {
            "checker": [MockFunction],
          },
        },
        "delete": Object {
          "dynamicCheckerReject": DynamicChecker {
            "checker": [MockFunction],
          },
        },
        "read": Object {
          "dynamicCheckerResolve": DynamicChecker {
            "checker": [MockFunction],
          },
        },
        "update": Object {
          "dynamicCheckerReject": DynamicChecker {
            "checker": [MockFunction],
          },
        },
      },
      "resource": Resource {
        "accessControl": [Circular],
        "children": Object {},
        "name": "resource2",
        "parent": Resource {
          "accessControl": [Circular],
          "children": Object {
            "resource1": Resource {
              "accessControl": [Circular],
              "children": Object {},
              "name": "resource1",
              "parent": [Circular],
              "path": "resource1",
            },
            "resource2": [Circular],
          },
          "name": "",
          "parent": null,
          "path": "",
        },
        "path": "resource2",
      },
      "role": Role {
        "accessControl": [Circular],
        "name": "role1",
      },
      "staticChecker": StaticChecker {
        "permissions": Object {
          "create": true,
          "delete": false,
          "read": false,
          "update": true,
        },
      },
    },
    "resource2#role2": Permission {
      "dynamicCheckers": Object {
        "create": Object {
          "dynamicCheckerResolve": DynamicChecker {
            "checker": [MockFunction],
          },
        },
        "delete": Object {
          "dynamicCheckerReject": DynamicChecker {
            "checker": [MockFunction],
          },
        },
        "read": Object {
          "dynamicCheckerResolve": DynamicChecker {
            "checker": [MockFunction],
          },
        },
        "update": Object {
          "dynamicCheckerReject": DynamicChecker {
            "checker": [MockFunction],
          },
        },
      },
      "resource": Resource {
        "accessControl": [Circular],
        "children": Object {},
        "name": "resource2",
        "parent": Resource {
          "accessControl": [Circular],
          "children": Object {
            "resource1": Resource {
              "accessControl": [Circular],
              "children": Object {},
              "name": "resource1",
              "parent": [Circular],
              "path": "resource1",
            },
            "resource2": [Circular],
          },
          "name": "",
          "parent": null,
          "path": "",
        },
        "path": "resource2",
      },
      "role": Role {
        "accessControl": [Circular],
        "name": "role2",
      },
      "staticChecker": StaticChecker {
        "permissions": Object {
          "create": true,
          "delete": false,
          "read": false,
          "update": true,
        },
      },
    },
  },
  "_resources": Resource {
    "accessControl": [Circular],
    "children": Object {
      "resource1": Resource {
        "accessControl": [Circular],
        "children": Object {},
        "name": "resource1",
        "parent": [Circular],
        "path": "resource1",
      },
      "resource2": Resource {
        "accessControl": [Circular],
        "children": Object {},
        "name": "resource2",
        "parent": [Circular],
        "path": "resource2",
      },
    },
    "name": "",
    "parent": null,
    "path": "",
  },
  "_roles": Object {
    "role1": Role {
      "accessControl": [Circular],
      "name": "role1",
    },
    "role2": Role {
      "accessControl": [Circular],
      "name": "role2",
    },
  },
}
`;

exports[`Access Control imports simplified dump 1`] = `
AccessControl {
  "_permissions": Object {
    "resource1#role1": Permission {
      "dynamicCheckers": Object {
        "create": Object {
          "dynamicCheckerResolve": DynamicChecker {
            "checker": [MockFunction],
          },
        },
        "delete": Object {},
        "read": Object {},
        "update": Object {
          "dynamicCheckerReject": DynamicChecker {
            "checker": [MockFunction],
          },
        },
      },
      "resource": Resource {
        "accessControl": [Circular],
        "children": Object {},
        "name": "resource1",
        "parent": Resource {
          "accessControl": [Circular],
          "children": Object {
            "resource1": [Circular],
            "resource2": Resource {
              "accessControl": [Circular],
              "children": Object {},
              "name": "resource2",
              "parent": [Circular],
              "path": "resource2",
            },
          },
          "name": "",
          "parent": null,
          "path": "",
        },
        "path": "resource1",
      },
      "role": Role {
        "accessControl": [Circular],
        "name": "role1",
      },
      "staticChecker": StaticChecker {
        "permissions": Object {
          "create": true,
          "delete": false,
          "read": false,
          "update": true,
        },
      },
    },
    "resource2#role1": Permission {
      "dynamicCheckers": Object {
        "create": Object {
          "dynamicCheckerResolve": DynamicChecker {
            "checker": [MockFunction],
          },
        },
        "delete": Object {},
        "read": Object {},
        "update": Object {
          "dynamicCheckerReject": DynamicChecker {
            "checker": [MockFunction],
          },
        },
      },
      "resource": Resource {
        "accessControl": [Circular],
        "children": Object {},
        "name": "resource2",
        "parent": Resource {
          "accessControl": [Circular],
          "children": Object {
            "resource1": Resource {
              "accessControl": [Circular],
              "children": Object {},
              "name": "resource1",
              "parent": [Circular],
              "path": "resource1",
            },
            "resource2": [Circular],
          },
          "name": "",
          "parent": null,
          "path": "",
        },
        "path": "resource2",
      },
      "role": Role {
        "accessControl": [Circular],
        "name": "role1",
      },
      "staticChecker": StaticChecker {
        "permissions": Object {
          "create": true,
          "delete": false,
          "read": false,
          "update": true,
        },
      },
    },
    "resource2#role2": Permission {
      "dynamicCheckers": Object {
        "create": Object {
          "dynamicCheckerResolve": DynamicChecker {
            "checker": [MockFunction],
          },
        },
        "delete": Object {},
        "read": Object {},
        "update": Object {
          "dynamicCheckerReject": DynamicChecker {
            "checker": [MockFunction],
          },
        },
      },
      "resource": Resource {
        "accessControl": [Circular],
        "children": Object {},
        "name": "resource2",
        "parent": Resource {
          "accessControl": [Circular],
          "children": Object {
            "resource1": Resource {
              "accessControl": [Circular],
              "children": Object {},
              "name": "resource1",
              "parent": [Circular],
              "path": "resource1",
            },
            "resource2": [Circular],
          },
          "name": "",
          "parent": null,
          "path": "",
        },
        "path": "resource2",
      },
      "role": Role {
        "accessControl": [Circular],
        "name": "role2",
      },
      "staticChecker": StaticChecker {
        "permissions": Object {
          "create": true,
          "delete": false,
          "read": false,
          "update": true,
        },
      },
    },
  },
  "_resources": Resource {
    "accessControl": [Circular],
    "children": Object {
      "resource1": Resource {
        "accessControl": [Circular],
        "children": Object {},
        "name": "resource1",
        "parent": [Circular],
        "path": "resource1",
      },
      "resource2": Resource {
        "accessControl": [Circular],
        "children": Object {},
        "name": "resource2",
        "parent": [Circular],
        "path": "resource2",
      },
    },
    "name": "",
    "parent": null,
    "path": "",
  },
  "_roles": Object {
    "role1": Role {
      "accessControl": [Circular],
      "name": "role1",
    },
    "role2": Role {
      "accessControl": [Circular],
      "name": "role2",
    },
  },
}
`;

exports[`Access Control returns resources 1`] = `
Resource {
  "accessControl": AccessControl {
    "_permissions": Object {
      "resource1#role1": Permission {
        "dynamicCheckers": Object {
          "create": Object {
            "dynamicCheckerResolve": DynamicChecker {
              "checker": [MockFunction],
            },
          },
          "delete": Object {},
          "read": Object {},
          "update": Object {
            "dynamicCheckerReject": DynamicChecker {
              "checker": [MockFunction],
            },
          },
        },
        "resource": Resource {
          "accessControl": [Circular],
          "children": Object {},
          "name": "resource1",
          "parent": [Circular],
          "path": "resource1",
        },
        "role": Role {
          "accessControl": [Circular],
          "name": "role1",
        },
        "staticChecker": StaticChecker {
          "permissions": Object {
            "create": true,
            "delete": false,
            "read": false,
            "update": true,
          },
        },
      },
      "resource2#role1": Permission {
        "dynamicCheckers": Object {
          "create": Object {
            "dynamicCheckerResolve": DynamicChecker {
              "checker": [MockFunction],
            },
          },
          "delete": Object {},
          "read": Object {},
          "update": Object {
            "dynamicCheckerReject": DynamicChecker {
              "checker": [MockFunction],
            },
          },
        },
        "resource": Resource {
          "accessControl": [Circular],
          "children": Object {},
          "name": "resource2",
          "parent": [Circular],
          "path": "resource2",
        },
        "role": Role {
          "accessControl": [Circular],
          "name": "role1",
        },
        "staticChecker": StaticChecker {
          "permissions": Object {
            "create": true,
            "delete": false,
            "read": false,
            "update": true,
          },
        },
      },
      "resource2#role2": Permission {
        "dynamicCheckers": Object {
          "create": Object {
            "dynamicCheckerResolve": DynamicChecker {
              "checker": [MockFunction],
            },
          },
          "delete": Object {},
          "read": Object {},
          "update": Object {
            "dynamicCheckerReject": DynamicChecker {
              "checker": [MockFunction],
            },
          },
        },
        "resource": Resource {
          "accessControl": [Circular],
          "children": Object {},
          "name": "resource2",
          "parent": [Circular],
          "path": "resource2",
        },
        "role": Role {
          "accessControl": [Circular],
          "name": "role2",
        },
        "staticChecker": StaticChecker {
          "permissions": Object {
            "create": true,
            "delete": false,
            "read": false,
            "update": true,
          },
        },
      },
    },
    "_resources": [Circular],
    "_roles": Object {
      "role1": Role {
        "accessControl": [Circular],
        "name": "role1",
      },
      "role2": Role {
        "accessControl": [Circular],
        "name": "role2",
      },
    },
  },
  "children": Object {
    "resource1": Resource {
      "accessControl": AccessControl {
        "_permissions": Object {
          "resource1#role1": Permission {
            "dynamicCheckers": Object {
              "create": Object {
                "dynamicCheckerResolve": DynamicChecker {
                  "checker": [MockFunction],
                },
              },
              "delete": Object {},
              "read": Object {},
              "update": Object {
                "dynamicCheckerReject": DynamicChecker {
                  "checker": [MockFunction],
                },
              },
            },
            "resource": [Circular],
            "role": Role {
              "accessControl": [Circular],
              "name": "role1",
            },
            "staticChecker": StaticChecker {
              "permissions": Object {
                "create": true,
                "delete": false,
                "read": false,
                "update": true,
              },
            },
          },
          "resource2#role1": Permission {
            "dynamicCheckers": Object {
              "create": Object {
                "dynamicCheckerResolve": DynamicChecker {
                  "checker": [MockFunction],
                },
              },
              "delete": Object {},
              "read": Object {},
              "update": Object {
                "dynamicCheckerReject": DynamicChecker {
                  "checker": [MockFunction],
                },
              },
            },
            "resource": Resource {
              "accessControl": [Circular],
              "children": Object {},
              "name": "resource2",
              "parent": [Circular],
              "path": "resource2",
            },
            "role": Role {
              "accessControl": [Circular],
              "name": "role1",
            },
            "staticChecker": StaticChecker {
              "permissions": Object {
                "create": true,
                "delete": false,
                "read": false,
                "update": true,
              },
            },
          },
          "resource2#role2": Permission {
            "dynamicCheckers": Object {
              "create": Object {
                "dynamicCheckerResolve": DynamicChecker {
                  "checker": [MockFunction],
                },
              },
              "delete": Object {},
              "read": Object {},
              "update": Object {
                "dynamicCheckerReject": DynamicChecker {
                  "checker": [MockFunction],
                },
              },
            },
            "resource": Resource {
              "accessControl": [Circular],
              "children": Object {},
              "name": "resource2",
              "parent": [Circular],
              "path": "resource2",
            },
            "role": Role {
              "accessControl": [Circular],
              "name": "role2",
            },
            "staticChecker": StaticChecker {
              "permissions": Object {
                "create": true,
                "delete": false,
                "read": false,
                "update": true,
              },
            },
          },
        },
        "_resources": [Circular],
        "_roles": Object {
          "role1": Role {
            "accessControl": [Circular],
            "name": "role1",
          },
          "role2": Role {
            "accessControl": [Circular],
            "name": "role2",
          },
        },
      },
      "children": Object {},
      "name": "resource1",
      "parent": [Circular],
      "path": "resource1",
    },
    "resource2": Resource {
      "accessControl": AccessControl {
        "_permissions": Object {
          "resource1#role1": Permission {
            "dynamicCheckers": Object {
              "create": Object {
                "dynamicCheckerResolve": DynamicChecker {
                  "checker": [MockFunction],
                },
              },
              "delete": Object {},
              "read": Object {},
              "update": Object {
                "dynamicCheckerReject": DynamicChecker {
                  "checker": [MockFunction],
                },
              },
            },
            "resource": Resource {
              "accessControl": [Circular],
              "children": Object {},
              "name": "resource1",
              "parent": [Circular],
              "path": "resource1",
            },
            "role": Role {
              "accessControl": [Circular],
              "name": "role1",
            },
            "staticChecker": StaticChecker {
              "permissions": Object {
                "create": true,
                "delete": false,
                "read": false,
                "update": true,
              },
            },
          },
          "resource2#role1": Permission {
            "dynamicCheckers": Object {
              "create": Object {
                "dynamicCheckerResolve": DynamicChecker {
                  "checker": [MockFunction],
                },
              },
              "delete": Object {},
              "read": Object {},
              "update": Object {
                "dynamicCheckerReject": DynamicChecker {
                  "checker": [MockFunction],
                },
              },
            },
            "resource": [Circular],
            "role": Role {
              "accessControl": [Circular],
              "name": "role1",
            },
            "staticChecker": StaticChecker {
              "permissions": Object {
                "create": true,
                "delete": false,
                "read": false,
                "update": true,
              },
            },
          },
          "resource2#role2": Permission {
            "dynamicCheckers": Object {
              "create": Object {
                "dynamicCheckerResolve": DynamicChecker {
                  "checker": [MockFunction],
                },
              },
              "delete": Object {},
              "read": Object {},
              "update": Object {
                "dynamicCheckerReject": DynamicChecker {
                  "checker": [MockFunction],
                },
              },
            },
            "resource": [Circular],
            "role": Role {
              "accessControl": [Circular],
              "name": "role2",
            },
            "staticChecker": StaticChecker {
              "permissions": Object {
                "create": true,
                "delete": false,
                "read": false,
                "update": true,
              },
            },
          },
        },
        "_resources": [Circular],
        "_roles": Object {
          "role1": Role {
            "accessControl": [Circular],
            "name": "role1",
          },
          "role2": Role {
            "accessControl": [Circular],
            "name": "role2",
          },
        },
      },
      "children": Object {},
      "name": "resource2",
      "parent": [Circular],
      "path": "resource2",
    },
  },
  "name": "",
  "parent": null,
  "path": "",
}
`;

exports[`Access Control returns roles 1`] = `
Object {
  "role1": Role {
    "accessControl": AccessControl {
      "_permissions": Object {
        "resource1#role1": Permission {
          "dynamicCheckers": Object {
            "create": Object {
              "dynamicCheckerResolve": DynamicChecker {
                "checker": [MockFunction],
              },
            },
            "delete": Object {},
            "read": Object {},
            "update": Object {
              "dynamicCheckerReject": DynamicChecker {
                "checker": [MockFunction],
              },
            },
          },
          "resource": Resource {
            "accessControl": [Circular],
            "children": Object {},
            "name": "resource1",
            "parent": Resource {
              "accessControl": [Circular],
              "children": Object {
                "resource1": [Circular],
                "resource2": Resource {
                  "accessControl": [Circular],
                  "children": Object {},
                  "name": "resource2",
                  "parent": [Circular],
                  "path": "resource2",
                },
              },
              "name": "",
              "parent": null,
              "path": "",
            },
            "path": "resource1",
          },
          "role": [Circular],
          "staticChecker": StaticChecker {
            "permissions": Object {
              "create": true,
              "delete": false,
              "read": false,
              "update": true,
            },
          },
        },
        "resource2#role1": Permission {
          "dynamicCheckers": Object {
            "create": Object {
              "dynamicCheckerResolve": DynamicChecker {
                "checker": [MockFunction],
              },
            },
            "delete": Object {},
            "read": Object {},
            "update": Object {
              "dynamicCheckerReject": DynamicChecker {
                "checker": [MockFunction],
              },
            },
          },
          "resource": Resource {
            "accessControl": [Circular],
            "children": Object {},
            "name": "resource2",
            "parent": Resource {
              "accessControl": [Circular],
              "children": Object {
                "resource1": Resource {
                  "accessControl": [Circular],
                  "children": Object {},
                  "name": "resource1",
                  "parent": [Circular],
                  "path": "resource1",
                },
                "resource2": [Circular],
              },
              "name": "",
              "parent": null,
              "path": "",
            },
            "path": "resource2",
          },
          "role": [Circular],
          "staticChecker": StaticChecker {
            "permissions": Object {
              "create": true,
              "delete": false,
              "read": false,
              "update": true,
            },
          },
        },
        "resource2#role2": Permission {
          "dynamicCheckers": Object {
            "create": Object {
              "dynamicCheckerResolve": DynamicChecker {
                "checker": [MockFunction],
              },
            },
            "delete": Object {},
            "read": Object {},
            "update": Object {
              "dynamicCheckerReject": DynamicChecker {
                "checker": [MockFunction],
              },
            },
          },
          "resource": Resource {
            "accessControl": [Circular],
            "children": Object {},
            "name": "resource2",
            "parent": Resource {
              "accessControl": [Circular],
              "children": Object {
                "resource1": Resource {
                  "accessControl": [Circular],
                  "children": Object {},
                  "name": "resource1",
                  "parent": [Circular],
                  "path": "resource1",
                },
                "resource2": [Circular],
              },
              "name": "",
              "parent": null,
              "path": "",
            },
            "path": "resource2",
          },
          "role": Role {
            "accessControl": [Circular],
            "name": "role2",
          },
          "staticChecker": StaticChecker {
            "permissions": Object {
              "create": true,
              "delete": false,
              "read": false,
              "update": true,
            },
          },
        },
      },
      "_resources": Resource {
        "accessControl": [Circular],
        "children": Object {
          "resource1": Resource {
            "accessControl": [Circular],
            "children": Object {},
            "name": "resource1",
            "parent": [Circular],
            "path": "resource1",
          },
          "resource2": Resource {
            "accessControl": [Circular],
            "children": Object {},
            "name": "resource2",
            "parent": [Circular],
            "path": "resource2",
          },
        },
        "name": "",
        "parent": null,
        "path": "",
      },
      "_roles": [Circular],
    },
    "name": "role1",
  },
  "role2": Role {
    "accessControl": AccessControl {
      "_permissions": Object {
        "resource1#role1": Permission {
          "dynamicCheckers": Object {
            "create": Object {
              "dynamicCheckerResolve": DynamicChecker {
                "checker": [MockFunction],
              },
            },
            "delete": Object {},
            "read": Object {},
            "update": Object {
              "dynamicCheckerReject": DynamicChecker {
                "checker": [MockFunction],
              },
            },
          },
          "resource": Resource {
            "accessControl": [Circular],
            "children": Object {},
            "name": "resource1",
            "parent": Resource {
              "accessControl": [Circular],
              "children": Object {
                "resource1": [Circular],
                "resource2": Resource {
                  "accessControl": [Circular],
                  "children": Object {},
                  "name": "resource2",
                  "parent": [Circular],
                  "path": "resource2",
                },
              },
              "name": "",
              "parent": null,
              "path": "",
            },
            "path": "resource1",
          },
          "role": Role {
            "accessControl": [Circular],
            "name": "role1",
          },
          "staticChecker": StaticChecker {
            "permissions": Object {
              "create": true,
              "delete": false,
              "read": false,
              "update": true,
            },
          },
        },
        "resource2#role1": Permission {
          "dynamicCheckers": Object {
            "create": Object {
              "dynamicCheckerResolve": DynamicChecker {
                "checker": [MockFunction],
              },
            },
            "delete": Object {},
            "read": Object {},
            "update": Object {
              "dynamicCheckerReject": DynamicChecker {
                "checker": [MockFunction],
              },
            },
          },
          "resource": Resource {
            "accessControl": [Circular],
            "children": Object {},
            "name": "resource2",
            "parent": Resource {
              "accessControl": [Circular],
              "children": Object {
                "resource1": Resource {
                  "accessControl": [Circular],
                  "children": Object {},
                  "name": "resource1",
                  "parent": [Circular],
                  "path": "resource1",
                },
                "resource2": [Circular],
              },
              "name": "",
              "parent": null,
              "path": "",
            },
            "path": "resource2",
          },
          "role": Role {
            "accessControl": [Circular],
            "name": "role1",
          },
          "staticChecker": StaticChecker {
            "permissions": Object {
              "create": true,
              "delete": false,
              "read": false,
              "update": true,
            },
          },
        },
        "resource2#role2": Permission {
          "dynamicCheckers": Object {
            "create": Object {
              "dynamicCheckerResolve": DynamicChecker {
                "checker": [MockFunction],
              },
            },
            "delete": Object {},
            "read": Object {},
            "update": Object {
              "dynamicCheckerReject": DynamicChecker {
                "checker": [MockFunction],
              },
            },
          },
          "resource": Resource {
            "accessControl": [Circular],
            "children": Object {},
            "name": "resource2",
            "parent": Resource {
              "accessControl": [Circular],
              "children": Object {
                "resource1": Resource {
                  "accessControl": [Circular],
                  "children": Object {},
                  "name": "resource1",
                  "parent": [Circular],
                  "path": "resource1",
                },
                "resource2": [Circular],
              },
              "name": "",
              "parent": null,
              "path": "",
            },
            "path": "resource2",
          },
          "role": [Circular],
          "staticChecker": StaticChecker {
            "permissions": Object {
              "create": true,
              "delete": false,
              "read": false,
              "update": true,
            },
          },
        },
      },
      "_resources": Resource {
        "accessControl": [Circular],
        "children": Object {
          "resource1": Resource {
            "accessControl": [Circular],
            "children": Object {},
            "name": "resource1",
            "parent": [Circular],
            "path": "resource1",
          },
          "resource2": Resource {
            "accessControl": [Circular],
            "children": Object {},
            "name": "resource2",
            "parent": [Circular],
            "path": "resource2",
          },
        },
        "name": "",
        "parent": null,
        "path": "",
      },
      "_roles": [Circular],
    },
    "name": "role2",
  },
}
`;
